src/io/GeoJSONReader.cpp

```diff
@@ -0,0 +1,269 @@
+    auto featuresJson = j["features"];
+    std::vector<GeoJSONFeature> features;
+    for(auto featureJson : featuresJson) {
+        auto f = readFeature(featureJson);
+        features.push_back(f);
+    }
+    return GeoJSONFeatureCollection{features};
+}
+
+
+std::unique_ptr<geom::Geometry> GeoJSONReader::readGeometry(nlohmann::json& j) {
+    std::string type = j["type"];
+    if (type == "Point") {
+        return readPoint(j);
+    } else if (type == "LineString") {
+        return readLineString(j);
+    } else if (type == "Polygon") {
+        return readPolygon(j);
+    } else if (type == "MultiPoint") {
+        return readMultiPoint(j);
+    } else if (type == "MultiLineString") {
+        return readMultiLineString(j);
+    } else if (type == "MultiPolygon") {
+        return readMultiPolygon(j);
+    } else if (type == "GeometryCollection") {
+        return readGeometryCollection(j);
+    }
+    return std::unique_ptr<geom::Geometry>(geometryFactory.createEmptyGeometry());
+}
+
+std::unique_ptr<geom::Point> GeoJSONReader::readPoint(nlohmann::json& j) {
+    std::vector<double> coords = j["coordinates"].get<std::vector<double>>();
+    if (coords.size() == 0) {
+        return std::unique_ptr<geom::Point>(geometryFactory.createPoint(2));
+    } else {
+        geom::Coordinate coord = {coords[0], coords[1]};
+        return std::unique_ptr<geom::Point>(geometryFactory.createPoint(coord));
+    }
+}
+
+std::unique_ptr<geom::LineString> GeoJSONReader::readLineString(nlohmann::json& j) {
+    std::vector<std::pair<double,double>> coords = j["coordinates"].get<std::vector<std::pair<double,double>>>();
+    std::vector<geom::Coordinate> coordinates;
+    for(std::vector<int>::size_type i = 0; i < coords.size(); i++) {
+        coordinates.push_back(geom::Coordinate{coords[i].first, coords[i].second});
+    }
+    geom::CoordinateArraySequence coordinateSequence { std::move(coordinates) };
+    return std::unique_ptr<geom::LineString>(geometryFactory.createLineString(coordinateSequence));
+}
+
+std::unique_ptr<geom::Polygon> GeoJSONReader::readPolygon(nlohmann::json& json) {
+    std::vector<std::vector<std::pair<double,double>>> polygonCoords = json["coordinates"].get<std::vector<std::vector<std::pair<double,double>>>>();
+    std::vector<geom::LinearRing *> rings;
+    for(std::vector<int>::size_type i = 0; i < polygonCoords.size(); i++) {
+        std::vector<geom::Coordinate> coordinates;
+        for (std::vector<int>::size_type j = 0; j < polygonCoords[i].size(); j++) {
+            coordinates.push_back(geom::Coordinate{polygonCoords[i][j].first, polygonCoords[i][j].second});
+        }
+        geom::CoordinateArraySequence coordinateSequence { std::move(coordinates) };
+        rings.push_back(geometryFactory.createLinearRing(std::move(coordinateSequence)));
+    }
+    if (rings.size() == 0) {
+        return std::unique_ptr<geom::Polygon>(geometryFactory.createPolygon(2));
+    } else if (rings.size() == 1) {
```

Comment:
What about one ring with zero coordinates in it? Should also return POLYGON EMPTY.

Reply:
I added unit test 21.